File: 프로그래머스/2/154538. 숫자 변환하기/숫자 변환하기.py
import math

def solution(x, y, n):
    d = [math.inf] * (y + 1)
    d[x] = 0
    x_mul_2 = x * 2
    x_mul_3 = x * 3
    
    for i in range(x + 1, y + 1):
        if (0 <= i - n):
            d[i] = min(d[i], d[i - n] + 1)
        
        if (i >= x_mul_2) and (i % 2 == 0):
            d[i] = min(d[i], d[i // 2] + 1)
            
        if (i >= x_mul_3) and (i % 3 == 0):
            d[i] = min(d[i], d[i // 3] + 1)
            
    return d[y] if d[y] != math.inf else -1

File: 프로그래머스/2/154538. 숫자 변환하기/숫자 변환하기.js
function solution(x, y, n) {
    var answer = 0;
    const MAX_VALUE = 1000001;
    const dp = Array(y + 1).fill(MAX_VALUE);
    dp[x] = 0;
    
    for (let i = x + 1; i <= y; i++) {
        // x에 n을 더하는 경우
        if (i - n >= x) dp[i] = Math.min(dp[i], dp[i - n] + 1);
        
        // x에 2를 곱하는 경우
        if (i % 2 === 0) dp[i] = Math.min(dp[i], dp[i / 2] + 1);
        
        // x에 3를 곱하는 경우
        if (i % 3 === 0) dp[i] = Math.min(dp[i], dp[i / 3] + 1);
        
    }
    
    return dp[y] === MAX_VALUE ? -1 : dp[y];
}